### PR TITLE
Integrate WorldViewer renderer

### PR DESCRIPTION
**Pull request by msteiger**
This PR integrates WorldViewer's core rendering functionality. This is what it looks like:

image

What it does in more detail:

* Add permissions for lambda expressions
* Move the preview by @Josharias into an implementation of `PreviewGenerator`
* Add blending operations for different color models (ARGB and RGBA for now)
* Renderers (aka `FacetLayer`s) for all facets in engine and Core.
* Tooltip information for all of them 

This also enables modules to add customized renderers for their own facets.

I use `BufferedImage` and `Graphics2D` for the rendering as I figured it would make most sense for 2D drawing operations (dashed lines, circles, polygons, etc). Terasology uses RGBA images and passes the backing byte buffer straight to OpenGL while WorldViewers draws into the default ARGB images.

Corresponding changes in WorldViewer have been made and will be pushed as soon as this is merged.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/68/
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Very nice :-)

On a related note if you're hacking at the preview part of the code could you check out #1524 ? Not related to this PR other than me wondering if previewing other world gens will still have the potential to cause crashes. We're not catching something we should be hardening against. I figure TTA might be missing some info to make a preview but that should just gracefully show an error rather than crash.

**Comment by msteiger** [comment]
Sure thing! I have a few other things pending (parallel computation, integration of configuration, etc) and will look into that in case the bug survives the refactoring.

**Comment by Josharias** [comment]
So cool!  Haven't been able to try it out yet.  Looks pretty spiffy, msteiger!  So much more awesome than my implementation.

File: engine/src/main/java/org/terasology/engine/module/ModuleManagerImpl.java [hunk 1]
```diff
@@ -106,6 +107,7 @@ private void setupSandbox() {
         moduleSecurityManager.getBasePermissionSet().addAPIPackage("java.util.concurrent");
         moduleSecurityManager.getBasePermissionSet().addAPIPackage("java.util.concurrent.atomic");
         moduleSecurityManager.getBasePermissionSet().addAPIPackage("java.util.concurrent.locks");
+        moduleSecurityManager.getBasePermissionSet().addAPIPackage("java.util.function");
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/engine/module/ModuleManagerImpl.java]: Anything scary in there? I figure not, just sounds sort of fundamental :-)

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/PreviewWorldScreen.java [hunk 2]
```diff
@@ -224,18 +193,19 @@ private void updatePreview() {
...
 
-        final ByteBufferProgressListener progressListener = new ByteBufferProgressListener() {
-            @Override
-            public void onProgress(float progress) {
+        ProgressListener progressListener = progress ->
                 popup.setMessage("Updating Preview", String.format("Please wait ... %d%%", (int) (progress * 100f)));
-            }
-        };
+
+        final int imageSize = 384;
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/PreviewWorldScreen.java]: Magic number? Maybe a comment or declaration up top? I see it used to be near class top

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/preview/FacetLayerPreview.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+import org.terasology.world.viewer.color.ColorModels;
+import org.terasology.world.viewer.layers.FacetLayer;
+import org.terasology.world.viewer.layers.FacetLayers;
+
+import com.google.common.math.IntMath;
+
+/**
+ * TODO Type description
+ * @author Martin Steiger
+ */
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/preview/FacetLayerPreview.java]: Outdated Javadoc template in IDE config I guess, we should fix :D

File: engine/src/main/java/org/terasology/world/viewer/color/Blender.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * limitations under the License.
+ */
+
+package org.terasology.world.viewer.color;
+
+
+/**
+ * Maps or blends color from a source into a target color space.
+ */
+public interface Blender {
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/world/viewer/color/Blender.java]: Maybe ambiguous class name? "ColorBlender" ?  With "Blender" also being a tool :-)

(nitpick, I admit)

File: engine/src/main/java/org/terasology/world/viewer/picker/package-info.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright 2013 MovingBlocks
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/world/viewer/picker/package-info.java]: Hardly a problem but it looks like we've got some outdated headers you're cloning

**Comment by Cervator** [comment]
Added some comments, mostly nitpicks, feel free to tweak or wait for some later pass :-)

Tests out fine locally and looks great, preview goes black for most alternative module worlds, what do module authors need to do so their worlds will show a nice preview?

TTA preview still crashes but the others just go black.

Feel free to merge when you like, I'm happy :-)

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/69/
Hooray Jenkins reported success with all tests good!

**Pull request merged**